File: algo/143. Reorder List.js
// Linked List

/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @return {void} Do not return anything, modify head in-place instead.
 */
var reorderList = function(head) {
    if (head === null) {
        return;
    }
    // find the middle
    var tortois = head;
    var rabbit = head;
    while (rabbit != null && rabbit.next != null) {
        tortois = tortois.next;
        rabbit = rabbit.next.next;
    }
    // cut the middle and reverse the second half
    var p1 = tortois;
    var p2 = tortois.next;
    tortois.next = null;
    while (p2 !== null) {
        var temp = p2.next;
        p2.next = p1;
        p1 = p2;
        p2 = temp;
    }
    // reorder the whole linked list
    cursor1 = head;
    cursor2 = p1;
    while (cursor2 !== null && cursor1.next !== cursor2) {
        var temp = cursor1.next;
        var temp2 = cursor2.next;
        cursor1.next = cursor2;
        cursor2.next = temp;
        cursor1 = cursor2.next;
        cursor2 = temp2;
    }
};

File: round3/143. Reorder List.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */
public class Solution {
    public void ReorderList(ListNode head) {
        // split the list into 2 with similar size
        ListNode dummy = new ListNode();
        dummy.next = head;
        ListNode slow = dummy, fast = dummy;
        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }      
        
        ListNode cursor2 = slow.next;
        slow.next = null;
        ListNode cursor1 = head;
        
        // reverse the 2nd list
        ListNode pre2 = null;
        while (cursor2 != null) {
            ListNode tmp = cursor2.next;
            cursor2.next = pre2;
            pre2 = cursor2;
            cursor2 = tmp;
        }
        cursor2 = pre2;
        
        // merge
        while (cursor1 != null && cursor2 != null) {
            ListNode tmp1 = cursor1.next, tmp2 = cursor2.next;
            cursor1.next = cursor2;
            cursor2.next = tmp1;
            cursor1 = tmp1;
            cursor2 = tmp2;
        }
        
    }
}